### PR TITLE
feat(backend): start_and_restart_docker_container_even_when_it_is_already_registered_in_Docker

### PR DESCRIPTION
**Pull request by alex1307**
Description

If container for a given tari image exists in Docker then it can't be started or restarted.

Motivation and Context

Launchpad should have full controll of tari containers and images.
Containers should be destroyed after shutdown, but it may happen to have abondened container.
On start we should check whether there is a docker container and destroy it.
Then we should create fresh new image and start it.

How Has This Been Tested?

Manually: Run Launchpad app, started and stopped containers.
Started containers from Launchpad and stopped manually on Docker.
Started the containers from Launchpad again.

### CODE REVIEW

File: applications/launchpad/backend/src/docker/mod.rs [hunk 1]
```diff
@@ -57,3 +62,90 @@ lazy_static! {
 lazy_static! {
     pub static ref CONTAINERS: RwLock<HashMap<String, ContainerState>> = RwLock::new(HashMap::new());
 }
+
+pub fn tari_blockchain_volume_name(tari_workspace: &str, tari_network: TariNetwork) -> String {
+    format!("{}_{}_volume", tari_workspace, tari_network.lower_case())
+}
+
+pub fn network_name(tari_workspace: &str) -> String {
```
Comment (CjS77) [inline applications/launchpad/backend/src/docker/mod.rs]: Confusingly named function.

This doesn't return the network name, but the key for this workspace's entry in a hashmap.

```suggestion
pub fn network_key(tari_workspace: &str) -> String {
```
```suggestion
pub fn network_name(tari_workspace: &str) -> String {
```

Comment (CjS77) [inline applications/launchpad/backend/src/docker/mod.rs]: Confusingly named function.

This doesn't return the network name, but the key for this workspace's entry in a hashmap.

```suggestion
pub fn network_key(tari_workspace: &str) -> String {
```
```suggestion
pub fn network_name(tari_workspace: &str) -> String {
```

Comment (CjS77) [reply applications/launchpad/backend/src/docker/mod.rs]: 1. Can this be a private function?
2. Can `tari_blockchain_volume_name` be private?

File: applications/launchpad/backend/src/docker/mod.rs [hunk 2]
```diff
@@ -57,3 +62,90 @@ lazy_static! {
...
+    let containers = &docker
+        .list_containers(Some(ListContainersOptions {
+            all: true,
+            filters: list_container_filters,
+            ..Default::default()
+        }))
+        .await?;
+
+    for c_id in containers.iter() {
+        let id = c_id.clone().id.unwrap();
```
Comment (CjS77) [inline applications/launchpad/backend/src/docker/mod.rs]: 1. Is this `clone` necessary? Looks like you can avoid it
2. Is the unwrap infallible?

File: applications/launchpad/backend/src/docker/settings.rs [hunk 3]
```diff
@@ -42,13 +42,13 @@ http://monero-stagenet.exan.tech:38081,\
 http://xmr-lux.boldsuck.org:38081,\
 http://singapore.node.xmr.pm:38081";
 
-#[derive(Default, Debug, Serialize, Deserialize)]
+#[derive(Default, Debug, Serialize, Deserialize, Clone)]
```
Comment (CjS77) [inline applications/launchpad/backend/src/docker/settings.rs]: If you don't actually clone these structs, don't bloat the codebase with the derived functions.

File: applications/launchpad/backend/src/docker/mod.rs [hunk 4]
```diff
@@ -57,3 +62,90 @@ lazy_static! {
...
+
+pub fn network_name(tari_workspace: &str) -> String {
+    format!("{}_network", tari_workspace)
+}
+
+pub async fn try_create_container(
+    image: ImageType,
+    fully_qualified_image_name: String,
+    tari_workspace: &str,
+    config: LaunchpadConfig,
```
Comment (CjS77) [inline applications/launchpad/backend/src/docker/mod.rs]: Pretty sure you can get away with a reference here

**Review by CjS77 — COMMENTED** [review]
Some minor issues. Generall y looks good.

File: applications/launchpad/backend/src/docker/settings.rs (hunk 3 above)
Comment (alex1307) [reply applications/launchpad/backend/src/docker/settings.rs]: pure java style :). Fixed

File: applications/launchpad/backend/src/docker/mod.rs (hunk 4 above)
Comment (alex1307) [reply applications/launchpad/backend/src/docker/mod.rs]: Changes are applied.

**Review by CjS77 — DISMISSED** [review]
LGTM

**Review by CjS77 — APPROVED** [review]
+1

**Pull request merged**